File: cdk/python/sample/more.py
"""The class for managing more environment

The class requires the follow properties:
    'id' (str): the suffix name of resource created
    'ec2_params' (dict): the dictionary of the EC2 custom parameters
    'lambda_params' (dict): the dictionary of the Lambda custom parameters

All properties are mandatory. See the unit tests for an example.
The class extendes the class named Basic.

# license MIT
# support https://github.com/bilardi/aws-simple-pipeline/issues
"""
from aws_cdk import (core, aws_ec2 as ec2,
                     aws_cloudwatch as cloudwatch)
from sample.basic import Basic
import json

class More(Basic):
    ec2_alarm = None
    lambda_alarm = None

    def __init__(self, scope: core.Construct, id: str, ec2_params: dict, lambda_params: dict, **kwargs) -> None:
        """
        deploys all AWS resources for more environment
            Resources:
                AWS::EC2::Instance with your details
                AWS::Lambda::Function with your policies
                AWS::Cloudwatch::Alarm for EC2 and Lambda
                AWS::Cloudwatch::Dashboard for EC2 and Lamnbda
        """
        super().__init__(scope, id, ec2_params, lambda_params, **kwargs)

        # ec2
        ec2_params['name'] = self.ec2_name
        self.ec2_alarm = self.get_alarm(ec2_params)
        ec2_dashboard = self.get_dashboard(ec2_params)

        # lambda
        lambda_params['name'] = self.lambda_name
        self.lambda_alarm = self.get_alarm(lambda_params)
        lambda_dashboard = self.get_dashboard(lambda_params)

    def get_vpc(self, ec2_params):
        vpc = None
        if 'vpc' in ec2_params:
            vpc = ec2_params['vpc']
        else:
            if 'vpc_id' in ec2_params and ec2_params['vpc_id']:
                vpc = ec2.Vpc.from_lookup(self, "vpc", vpc_id=ec2_params['vpc_id'])
            else:
                vpc = ec2.Vpc.from_lookup(self, "vpc", is_default=True)
        return vpc
    
    def get_security_group(self, ec2_params):
        security_group = None
        if 'security_group' in ec2_params:
            security_group = ec2_params['security_group']
        else:
            if 'security_group_id' in ec2_params and ec2_params['security_group_id']:
                security_group = ec2.SecurityGroup.from_security_group_id(self, "SecurityGroup",
                    security_group_id=ec2_params['security_group_id'],
                    mutable=False
                )
            else:
                security_group = ec2.SecurityGroup(self, "SecurityGroup",
                    vpc=ec2_params['vpc']
                )
                security_group.add_ingress_rule(
                    peer=ec2.Peer.any_ipv4(),
                    connection=ec2.Port(
                        string_representation="sr",
                        protocol=ec2.Protocol("UDP"),
                        from_port=ec2_params['from_port'],
                        to_port=ec2_params['to_port']
                    )
                )
        return security_group

    def get_block_device(self, params):
        volume = ec2.BlockDeviceVolume.ebs(
            delete_on_termination=params['delete_on_termination'],
            volume_size=params['volume_size'],
            volume_type=ec2.EbsDeviceVolumeType(params['volume_type'])
        )
        block_device = ec2.BlockDevice(
            device_name=params['device_name'],
            volume=volume
        )
        return block_device

    def get_instance(self, ec2_params):
        if 'vpc' not in ec2_params:
            ec2_params['vpc'] = self.get_vpc(ec2_params)
        if 'security_group' not in ec2_params:
            ec2_params['security_group'] = self.get_security_group(ec2_params)
        block_device = self.get_block_device(ec2_params)
        ec2_instance = ec2.Instance(self, self.ec2_name,
            machine_image=ec2.MachineImage.generic_linux(
                ami_map={ec2_params['region']:ec2_params['ami_id']}
            ),
            vpc=ec2_params['vpc'],
            security_group=ec2_params['security_group'],
            instance_type=ec2.InstanceType(ec2_params['instance_type']),
            key_name=ec2_params['key_name'],
            block_devices=[block_device]
        )
        ec2_instance.user_data.add_commands(ec2_params['user_data'])
        self.add_tags(self.ec2_name, ec2_instance, ec2_params['tags'])
        return ec2_instance

    def get_alarm(self, params):
        description = params['description'].format(params['name'], self.account)
        metric = cloudwatch.Metric(
            metric_name=params['metric_name'],
            namespace=params['namespace'],
            dimensions={params['dimension']:params['name']}
        )
        alarm = cloudwatch.Alarm(self, "{}Alarm".format(params['name']),
            alarm_description=description,
            alarm_name=description,
            comparison_operator=cloudwatch.ComparisonOperator('GREATER_THAN_OR_EQUAL_TO_THRESHOLD'),
            metric=metric,
            evaluation_periods=params['evaluation_periods'],
            period=core.Duration.seconds(params['period']),
            statistic=params['statistic'],
            threshold=params['threshold'],
            treat_missing_data=cloudwatch.TreatMissingData('MISSING')
        )
        return alarm

    def get_dashboard(self, params):
        with open(params['dashboard_file']) as json_file:
            params['dashboard_widgets'] = json.load(json_file)
        graph_widgets = []
        for widget in params['dashboard_widgets']:
            metric = [cloudwatch.Metric(
                namespace=widget['properties']['metrics'][0][0],
                metric_name=widget['properties']['metrics'][0][1],
                dimensions={widget['properties']['metrics'][0][2]: params['name']}
            )]
            graph_widget = cloudwatch.GraphWidget(
                height=widget['height'],
                width=widget['width'],
                left=metric
            )
            graph_widget.position(widget['x'], widget['y'])
            graph_widgets.append(graph_widget)
        dashboard = cloudwatch.Dashboard(self, "{}Dashboard".format(params['name']),
            dashboard_name=params['name'],
            widgets=[graph_widgets]
        )
        return dashboard        

File: cdk/typescript/lib/more.ts
/**
 * The class for managing more environment
 * 
 * The class requires some properties described in the constructor comment.
 * The class extendes the class named Basic.
 * 
 * The class deploys all AWS resources for more environment
 * 
 *   AWS::EC2::Instance with your details
 *   AWS::Lambda::Function with your policies
 *   AWS::Cloudwatch::Alarm for EC2 and Lambda
 *   AWS::Cloudwatch::Dashboard for EC2 and Lamnbda
 * 
 * **license**: MIT,
 * **support**: https://github.com/bilardi/aws-tool-comparison/issues
 */
import * as cdk from "@aws-cdk/core";
import * as ec2 from "@aws-cdk/aws-ec2";
import * as cloudwatch from "@aws-cdk/aws-cloudwatch";
import { Basic } from "./basic";
import ec2DashboardWidgets from "./ec2.dashboard.widgets.json";
import lambdaDashboardWidgets from "./lambda.dashboard.widgets.json";

export class More extends Basic {
  ec2Alarm:cloudwatch.Alarm;
  lambdaAlarm:cloudwatch.Alarm;

  /**
   * All properties are mandatory. See the unit tests for an example.
   * 
   * @param id the suffix name of resource created
   * @param ec2Params the dictionary of the EC2 custom parameters
   * @param lambdaParams the dictionary of the Lambda custom parameters
   */
  constructor(scope: cdk.Construct, id: string, ec2Params: {[key: string]: any}, lambdaParams: {[key: string]: any}, props?: cdk.StackProps) {
    super(scope, id, ec2Params, lambdaParams, props);

    // ec2
    ec2Params.name = this.ec2Name;
    ec2Params.dashboardWidgets = ec2DashboardWidgets;
    this.ec2Alarm = this.getAlarm(ec2Params);
    this.getDashboard(ec2Params);

    // lambda
    lambdaParams.name = this.lambdaName;
    lambdaParams.dashboardWidgets = lambdaDashboardWidgets;
    this.lambdaAlarm = this.getAlarm(lambdaParams);
    this.getDashboard(lambdaParams);
  }

  format(str: string, ...args: string[]): string {
    return str.replace(/{(\d+)}/g, (match, index) => args[index] || "");
  }

  getVpc(ec2Params: {[key: string]: any}): ec2.IVpc {
    let vpc = undefined;
    if ("vpc" in ec2Params) {
        vpc = ec2Params.vpc;
    } else {
      if ("vpcId" in ec2Params && ec2Params.vpcId) {
        vpc = ec2.Vpc.fromLookup(this, "vpc", { vpcId: ec2Params.vpcId });
      } else {
        vpc = ec2.Vpc.fromLookup(this, "vpc", { isDefault: true });
      }
    }
    return vpc;
  }

  getSecurityGroup(ec2Params: {[key: string]: any}): ec2.ISecurityGroup {
    let securityGroup = undefined;
    if ("securityGroup" in ec2Params) {
      securityGroup = ec2Params.securityGroup;
    } else {
      if ("securityGroupId" in ec2Params && ec2Params.securityGroupId) {
        securityGroup = ec2.SecurityGroup.fromSecurityGroupId(this, "SecurityGroup",
          ec2Params.securityGroupId,
          { mutable: false }
        );
      } else {
        securityGroup = new ec2.SecurityGroup(this, "SecurityGroup",
          { vpc: ec2Params.vpc }
        );
        securityGroup.addIngressRule(
          ec2.Peer.anyIpv4(),
          new ec2.Port({
            stringRepresentation: "sr",
            protocol: ec2.Protocol.UDP,
            fromPort: ec2Params.fromPort,
            toPort: ec2Params.toPort
          })
        );
      }
    }
    return securityGroup;
  }

  getBlockDevice(ec2Params: {[key: string]: any}): ec2.BlockDevice[] {
    const volume = ec2.BlockDeviceVolume.ebs(
      ec2Params.volumeSize,
      {
        deleteOnTermination: ec2Params.deleteOnTermination,
        volumeType: ec2Params.volumeType
      }
    );
    const blockDevices: ec2.BlockDevice[] = [
      {
        deviceName: ec2Params.name,
        volume: volume
      }
    ];
    return blockDevices;
  }

  getInstance(ec2Params: {[key: string]: any}): ec2.Instance {
    if (!("vpc" in ec2Params)) {
      ec2Params.vpc = this.getVpc(ec2Params);
    }
    if (!("securityGroup" in ec2Params)) {
      ec2Params.securityGroup = this.getSecurityGroup(ec2Params);
    }
    // const blockDevices = this.getBlockDevice(ec2Params);
    const ec2Instance = new ec2.Instance(this, this.ec2Name, {
      vpc: ec2Params.vpc,
      instanceType: new ec2.InstanceType(ec2Params.instanceType),
      machineImage: ec2.MachineImage.genericLinux(ec2Params.amiMap),
      securityGroup: ec2Params.securityGroup,
      keyName: ec2Params.keyName,
      // blockDevices: blockDevices
    });
    ec2Instance.addUserData(ec2Params.userData);
    this.addTags(this.ec2Name, ec2Instance, ec2Params.tags);
    return ec2Instance;
  }

  getAlarm(params: {[key: string]: any}): cloudwatch.Alarm {
    const description = this.format(params.description, params.name, this.account);
    const dimensions: {[key: string]: string} = {};
    dimensions[params.dimension] = params.name;
    const metric = new cloudwatch.Metric({
      metricName:params.metricName,
      namespace:params.namespace,
      dimensions:dimensions,
    });
    return new cloudwatch.Alarm(this, this.format("{0}Alarm", params.name), {
      alarmDescription:description,
      alarmName:description,
      comparisonOperator:cloudwatch.ComparisonOperator.GREATER_THAN_OR_EQUAL_TO_THRESHOLD,
      metric:metric,
      evaluationPeriods:params.evaluationPeriods,
      period:cdk.Duration.seconds(params.period),
      statistic:params.statistic,
      threshold:params.threshold,
      treatMissingData:cloudwatch.TreatMissingData.MISSING
    });
  }

  getDashboard(params: {[key: string]: any}): cloudwatch.Dashboard {
    const graphWidgets = [];
    for (const widget of params.dashboardWidgets) {
      const dimensions: {[key: string]: string} = {};
      dimensions[widget["properties"]["metrics"][0][2]] = params.name;
      const metric = [new cloudwatch.Metric({
        namespace:widget["properties"]["metrics"][0][0],
        metricName:widget["properties"]["metrics"][0][1],
        dimensions:dimensions
      })];
      const graphWidget = new cloudwatch.GraphWidget({
        height:widget["height"],
        width:widget["width"],
        left:metric
      });
      graphWidget.position(widget["x"], widget["y"]);
      graphWidgets.push(graphWidget);
    }
    return new cloudwatch.Dashboard(this, this.format("{0}Dashboard", params.name), {
      dashboardName:params.name,
      widgets:[graphWidgets]
    });
  }
}
